x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -853,75 +840,76 @@ private void putTransientIfNonExisting(String key, Object value) {
-            final String apiKeyId = (String) authentication.getMetadata().get(AuthenticationField.API_KEY_ID_KEY);
-            assert apiKeyId != null : "api key id must be present in the metadata";
-            userText = "API key id [" + apiKeyId + "] of " + userText;
-        }
+    private ElasticsearchSecurityException actionDenied(Authentication authentication, String action, TransportRequest request) {
+        return actionDenied(authentication, action, request, null);
+    }
 
-        // The run-as user is always from a realm. So it must have roles that can be printed.
-        // If the user is not run-as, we cannot print the roles if it's an API key or a service account (both do not have
-        // roles, but privileges)
-        if (false == authentication.isServiceAccount() && false == authentication.isApiKey()) {
-            userText = userText + " with roles [" + Strings.arrayToCommaDelimitedString(authentication.getUser().roles()) + "]";
-        }
+    private ElasticsearchSecurityException actionDenied(
+        Authentication authentication,
+        String action,
+        TransportRequest request,
+        Exception cause
+    ) {
+        return actionDenied(authentication, action, request, null, cause);
+    }
 
-        String message = "action [" + action + "] is unauthorized for " + userText;
-        if (context != null) {
-            message = message + " " + context;
-        }
+    private ElasticsearchSecurityException actionDenied(
+        Authentication authentication,
+        String action,
+        TransportRequest request,
+        @Nullable String context,
+        Exception cause
+    ) {
+        return denialException(
+            authentication,
+            action,
+            () -> AuthorizationDenialMessages.actionDenied(authentication, action, request, context),
+            cause
+        );
+    }
 
-        if (ClusterPrivilegeResolver.isClusterAction(action)) {
-            final Collection<String> privileges = ClusterPrivilegeResolver.findPrivilegesThatGrant(action, request, authentication);
-            if (privileges != null && privileges.size() > 0) {
-                message = message
-                    + ", this action is granted by the cluster privileges ["
-                    + collectionToCommaDelimitedString(privileges)
-                    + "]";
-            }
-        } else if (isIndexAction(action)) {
-            final Collection<String> privileges = IndexPrivilege.findPrivilegesThatGrant(action);
-            if (privileges != null && privileges.size() > 0) {
-                message = message
-                    + ", this action is granted by the index privileges ["
-                    + collectionToCommaDelimitedString(privileges)
-                    + "]";
-            }
+    private ElasticsearchSecurityException denialException(
+        Authentication authentication,
+        String action,
+        Supplier<String> messageSupplier,
```

Comment:
Do we really need `Supplier` here vs a simple `String`?

Reply:
I went with a supplier because `denialException` doesn't always use the passed in message. When we're in anonymous mode and authz exceptions are disabled, we return an authc exception instead. The supplier signals that a little better than passing in a string. I'm considering a rename to `authzDenialMessageSupplier` to make it more explicit. It also preserves the flow we had before this PR: we return an authc exception in the anonymous case, and construct an authz exception otherwise. As such prefer the supplier, despite the added indirection. 


wouldn't get constructed at all if we're in anonymous mode and authz exceptions are disabled. In terms of control flow this makes sense: if we need an authentication exception, we construct that, otherwise we construct an authz exception. 

Though it adds indirection in comparison to passing in a String, I'd prefer to preserve the flow and use a supplier. Wdyt?

Reply:
No problem. I have no strong opinion.